proposals/open-telemetry/microprofile-opentelemetry-tracing.asciidoc

```diff
@@ -0,0 +1,190 @@
+:authors: Roberto Cortez
+:email: radcortez@yahoo.com
+:version-label!:
+:sectanchors:
+:doctype: book
+:license: Eclipse Foundation Specification License v1.0
+:source-highlighter: coderay
+:toc: left
+:toclevels: 4
+:sectnumlevels: 4
+ifdef::backend-pdf[]
+:pagenums:
+endif::[]
+
+= Eclipse MicroProfile OpenTelemetry Tracing
+
+== Introduction
+
+TODO
+
+== Rationale
+
+TODO
+
+== Architecture
+
+This specification defines an easy way to allow an application running in a MicroProfile container to take advantage of
+distributed tracing by using the https://opentelemetry.io[OpenTelemetry project] (a merger of the OpenTracing and
+OpenCensus projects).
+
+This document and implementations *MUST* comply with the following OpenTelemetry specifications:
+
+- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/overview.md[OpenTelemetry Overview]
+(except for Metrics)
+- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md[Tracing API]
+- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/baggage/api.md[Baggage API]
+- https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/context[Context API]
+- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md[Resource SDK]
+
+=== Automatic Instrumentation
+
+JAX-RS applications, JAX-RS Clients, and MicroProfile REST Clients are automatically enlisted to participate in
+distributed tracing without code modification.
+
```

Comment:
Could you explain why we need to make this distinction?

Reply:
> Could you explain why we need to make this distinction?

The intent is that anything related to REST (and users can interact with REST through JAX-RS Server, JAX-RS Clients and MP REST Client) is automatically traced without additional code. I list all the options, so it is clear which places support the automatic instrumentation.

Reply:
> Suggested change 
> Correct parent child relationships between incoming requests and outgoing requests are handled automatically, as long as the outgoing requests occur on the same thread as the incoming request. If outgoing requests are performed on a different thread than the incoming request, it is the developers responsibility to propagate the Tracer context between threads.

I believe this is not required, because it is already covered by the OTel specification.

Reply:
> > Could you explain why we need to make this distinction?
> 
> The intent is that anything related to REST (and users can interact with REST through JAX-RS Server, JAX-RS Clients and MP REST Client) is automatically traced without additional code. I list all the options, so it is clear which places support the automatic instrumentation.

Sorry, @radcortez. That question was for @Emily-Jiang. :)

Reply:
I think it is nice for this clarification.

Reply:
In today's call, we agreed that adding the link to the OpenT spec is helpful.

Reply:
Which link? There is only text in this suggestion.

BTW, I believe this text is not correct (I know it was copied from the MP OpenTracing spec, and is not correct there either). What establishes parent-child relationships is the traceId sitting in the OTel Context. There are situations where you may have the same thread handling the request and response and the OTel Context being different.

In my opinion, we should just link to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span (which explains how the relationship between parent and child is established), but this is just a subsection of the Trace API already linked in the Architecture section.

Reply:
Added link to the Trace API in this section, which contains the rules to establish the parent - child relationships.